Add support for blockquotes coming from formatters

jgadbois opened the pull request:
This will require formatters to implement a blockquote method.

jgadbois committed:
Add code to blockquote based on formatter

jgadbois committed:
We want to quote the reply_to post

jgadbois committed:
Set text in controller

jgadbois committed:
Escape raw post text so that blockquotes get formatted

radar commented on app/helpers/forem/formatting_helper.rb:
```diff
@@ -8,5 +8,13 @@ module Forem
     end
+
+    def as_quoted_text(text)
+      if Forem.formatter
```
How about this instead?

```
if Forem.formatter && Forem.formatter.respond_to?(:blockquote)
```

That would mean that you wouldn't break existing formatters and future formatters wouldn't need to implement a blockquote method.

jgadbois replied:
Yeah, I like that - committed

jgadbois committed:
Don't break existing formatters

parndt commented:
Can we please have tests for this functionality?

Thanks!

jgadbois committed:
Added specs for blockquoting

jgadbois commented:
Started on some specs which brought up a problem - all the formatters run the raw markdown through ERB::Util.h which escapes the markdown blockquote formatting (start the line with greater than).

The current commit in the pull request gets around this by wrapping the post in raw() but that's probably not smart and causes an existing spec to fail.

I'm pretty sure the blockquote tag gets escaped using the current functionality too.

Any suggestions?  I was thinking there might be some markdown sanitizer we could use instead of h()?

parndt commented:
What do you think @radar ?

radar commented:
Ok, perhaps we should be running the formatter and then running that output through a HTML sanitizer. Any experience with this @parndt?

parndt commented:
@radar none

jgadbois committed:
Sanitize formatter output

jgadbois committed:
Add as_sanitized_text method

jgadbois commented:
I removed the escaping from the RDiscount gem and added a call to sanitize before outputting the forem post.  Specs (my new ones and existing ones) are passing.

jgadbois committed:
Merge remote-tracking branch 'upstream/master'

jgadbois committed:
Merge branch 'master' into formatted_quotes

jgadbois commented:
Sorry some of these pull request commits are on my master in addition to my formatted-quotes branch, so I'll have to have this pull request merged before I can fork off another branch to start working on the email subscriptions.  I'm pretty new to pull requests so sorry for messing that up.

jgadbois commented:
It should - I'm sanitizing before running through markdown.

radar commented:
Alright, I'll run through it tonight when I'm home from work.

jgadbois commented:
@radar - sorry to bug you.  Had a chance to check this out yet?

radar commented:
No sorry I've not had the time, yesterday's work took longer than I thought and I also spent the morning + evening biking 50km.  

I am in Adelaide this weekend (until the 27th) visiting my family. I will try to find some time to look into it this weekend. 

On Saturday, 17 December 2011 at 2:32 AM, John Gadbois wrote:

> @radar - sorry to bug you. Had a chance to check this out yet?
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/radar/forem/pull/79#issuecomment-3178507

radar commented:
There's a binary file in this called "core". What is that?

radar commented:
This seems to not work... when I click "Quote" on a post it gives me the `<blockquote>` wrapping for the text, but when the post is created this tag vanishes.

radar commented:
I've made [a new commit to forem-rdiscount here](https://github.com/radar/forem-rdiscount/commit/84470d4) which seems to fix this error.

jgadbois commented:
Thanks for committing this - I'll get started on email notices.
